internal/publish/metrics.go

```diff
@@ -0,0 +1,115 @@
+	mResponder := maintenance.New(h.config)
+	return h.tracker.HandleTrack(chaff.HeaderDetector("X-Chaff"),
+		mResponder.Handle(
+			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				ctx, span := trace.StartSpan(r.Context(), "(*publish.HandleMetrics)")
+				defer span.End()
+
+				var request verifyapi.MetricsRequest
+				code, err := jsonutil.Unmarshal(w, r, &request)
+				if err != nil {
+					message := fmt.Sprintf("error unmarshalling API call, code: %v: %v", code, err)
+					span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: message})
+					errorCode := verifyapi.ErrorBadRequest
+					if code == http.StatusInternalServerError {
+						errorCode = verifyapi.ErrorInternalError
+					}
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusBadRequest)
+					fmt.Fprintf(w, "{\"error\": \"%v\", \"code\": \"%v\"}", err.Error(), errorCode)
+					return
+				}
+
+				response, status := h.handleMetricsRequest(ctx, r.Header.Get("Authorization"), &request)
+
+				data, err := json.Marshal(response)
+				if err != nil {
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusInternalServerError)
+					fmt.Fprintf(w, "{\"error\": \"%v\"}", err.Error())
+					return
+				}
+
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(status)
+				fmt.Fprintf(w, "%s", data)
+			})))
+}
+
+func (h *PublishHandler) handleMetricsRequest(ctx context.Context, bearerToken string, request *verifyapi.MetricsRequest) (*verifyapi.MetricsResponse, int) {
+	logger := logging.FromContext(ctx)
+	response := &verifyapi.MetricsResponse{}
+
+	if !strings.HasPrefix(bearerToken, "Bearer ") {
+		response.ErrorMessage = "Authorization header is not in `Bearer <token>` format"
+		return response, http.StatusBadRequest
+	}
+	// Remove 'Bearer ' from the token.
+	bearerToken = bearerToken[7:]
+
+	// Validate JWT - if valid, the health authority ID (based on issuer) is returned.
+	healthAuthorityID, err := h.verifier.AuthenticateStatsToken(ctx, bearerToken)
+	if err != nil {
+		response.ErrorMessage = err.Error()
+		response.ErrorCode = verifyapi.ErrorUnauthorized
+		return response, http.StatusUnauthorized
+	}
+
+	// retrieve stats
+	stats, err := h.database.ReadStats(ctx, healthAuthorityID)
+	if err != nil {
+		logger.Errorw("error reading stats", "error", err)
+		response.ErrorMessage = "error reading stats"
+		response.ErrorCode = verifyapi.ErrorInternalError
+		return response, http.StatusInternalServerError
+	}
+
+	// Nothing from the current hour can be shown.
+	onlyBefore := time.Now().UTC().Truncate(time.Hour)
+
+	// Combine days - this also filters things that are "too new" and days that don't meet the threshold.
+	metricsDays := model.ReduceStats(stats, onlyBefore, h.config.StatsUploadMinimum)
+	// Transform array of points to array of structs before return.
+	response.Days = make([]verifyapi.MetricsDay, len(metricsDays))
```

Comment:
It'd be better to pass around `[]*verifyapi.MetricsDay` (pointers)

Reply:
this is intentionally - the response struct isn't pointers (perhaps it should be?)

the return is a pointer to the struct that contains this slice.